File: 350_intersection_of_two_arrays_II/intersect.py
from collections import defaultdict
from typing import List


class Solution:
    def intersect(self, nums1: List[int], nums2: List[int]) -> List[int]:
        if len(nums1) > len(nums2): return self.intersect(nums2, nums1)

        res = []

        h = defaultdict(int)

        for i in nums1:
            h[i] += 1

        for i in nums2:
            if i in h and h[i] > 0:
                res.append(i)
                h[i] -= 1

        return res

File: 350_intersection_of_two_arrays_II/intersect.rs
use std::collections::HashMap;

impl Solution {
    pub fn intersect(nums1: Vec<i32>, nums2: Vec<i32>) -> Vec<i32> {
        let mut h_map: HashMap<i32, i32> = HashMap::new();
        let mut res = Vec::new();

        for num in nums1 {
            h_map.entry(num)
                .and_modify(|i| {*i+=1})
                .or_insert(1);
        }

        for num in nums2 {
            h_map.entry(num)
                .and_modify(|i| {
                    if *i > 0 {
                        res.push(num);
                    }
                    *i-=1;
                });
        }

        res
    }
}